opm/simulators/wells/BlackoilWellModel.hpp

```diff
@@ -215,14 +215,113 @@ namespace Opm {
+            */
+
+
+            /*
+              Immutable version of the currently active wellstate.
+            */
+            const WellStateFullyImplicitBlackoil& wellState() const
+            {
+                return this->active_well_state_;
+            }
+
+            /*
+              Mutable version of the currently active wellstate.
+            */
+            WellStateFullyImplicitBlackoil& wellState()
+            {
+                return this->active_well_state_;
+            }
+
+            /*
+              Will return the last good wellstate. This is typcially used when
+              initializing a new report step where the Schedule object might
+              have introduced new wells. The wellstate returned by
+              prevWellState() must have been stored with the commitWellState()
+              function first.
+            */
+            const WellStateFullyImplicitBlackoil& prevWellState() const
+            {
+                return this->last_valid_well_state_;
+            }
+
+            /*
+              Will return the currently active nupcolWellState; must initialize
+              the internal nupcol wellstate with initNupcolWellState() first.
+            */
+            const WellStateFullyImplicitBlackoil& nupcolWellState() const
+            {
+                return this->nupcol_well_state_;
+            }
+
+            /*
+              Will assign the internal member last_valid_well_state_ to the
+              current value of the this->active_well_state_. The state stored
+              with storeWellState() can then subsequently be recovered with the
+              resetWellState() method.
+            */
+            void commitWellState()
+            {
+                this->last_valid_well_state_ = this->active_well_state_;
+            }
+
+            /*
+              Will store a copy of the input argument well_state in the
+              last_valid_well_state_ member, that state can then be recovered
+              with a subsequent call to resetWellState().
+            */
+            void commitWellState(WellStateFullyImplicitBlackoil well_state)
+            {
+                this->last_valid_well_state_ = std::move(well_state);
+            }
+
+            /*
+              Will update the internal variable active_well_state_ to whatever
+              was stored in the last_valid_well_state_ member. This function
+              works in pair with commitWellState() which should be called first.
+            */
+            void resetWellState()
+            {
+                this->active_well_state_ = this->last_valid_well_state_;
+            }
+
+            /*
+              Will store the current active wellstate in the nupcol_well_state_
+              member. This can then be subsequently retrieved with accessor
+              nupcolWellState().
+            */
+            void initNupcolWellState()
```

Comment:
I suggest the name updateNupcolWellState() instead, since this will be called for each of the first NUPCOL iterations.

Reply:
OK